### PR TITLE
Added the ellipse geometry

### PR DESCRIPTION
**Pull request by anranjiao**

### CODE REVIEW

**Comment by lululxvi** [comment]
Reformat via black.

File: deepxde/geometry/geometry_2d.py [hunk 1]
```diff
@@ -69,6 +69,55 @@ def background_points(self, x, dirn, dist2npt, shift):
         return pts
 
 
+class Ellipse(Geometry):
+    def __init__(self, center, long_axis, short_axis, alpha=0):
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:73]: Need to add a docstring to explain these arguments.

File: deepxde/geometry/geometry_2d.py [hunk 2]
```diff
@@ -65,10 +67,86 @@ def background_points(self, x, dirn, dist2npt, shift):
...
 
 
+class Ellipse(Geometry):
+    """Ellipse.
+
+    Args:
+        center: Center of the ellipse.
+        long_axis: Length of semi-major axis of the ellipse.
+        short_axis: Length of semi-minor axis of the ellipse.
+        alpha: Rotation angle of the ellipse.
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:86]: change `alpha` to `angle`. You need also to mention that the rotation is counterclockwise, see https://docs.plm.automation.siemens.com/data_services/resources/nx/11/nx_api/custom/en_US/grip/pt_curve/ptc_geom_ellipse.html

File: deepxde/geometry/geometry_2d.py [hunk 3]
```diff
@@ -65,39 +67,64 @@ def background_points(self, x, dirn, dist2npt, shift):
...
+        )
         return pts
 
 
 class Ellipse(Geometry):
+    """Ellipse.
+
+    Args:
+        center: Center of the ellipse.
+        long_axis: Length of semi-major axis of the ellipse.
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:84]: Use semimajor and semiminor

File: deepxde/geometry/geometry_2d.py [hunk 4]
```diff
@@ -65,39 +67,64 @@ def background_points(self, x, dirn, dist2npt, shift):
...
-                                     center[1] + self.radius * np.sin(alpha)], 
-                                     dtype=config.real(np))
-        self.right_center = np.array([center[0] + self.radius * np.cos(alpha), 
-                                      center[1] - self.radius * np.sin(alpha)], 
-                                      dtype=config.real(np))
-        self.rotate_coeff = np.array([[np.cos(-alpha), -np.sin(-alpha)],
-                                      [np.sin(-alpha),  np.cos(-alpha)]])
-        self._r2 = self.radius ** 2       
-        super().__init__(2, (self.center - long_axis, self.center + short_axis), 2 * self.radius)
+        self.radius = (long_axis**2 - short_axis**2) ** 0.5
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:94]: radius is not a good name. https://en.wikipedia.org/wiki/Ellipse

maybe self.c?

File: deepxde/geometry/geometry_2d.py [hunk 5]
```diff
@@ -65,10 +67,89 @@ def background_points(self, x, dirn, dist2npt, shift):
...
+        )
         return pts
 
 
+class Ellipse(Geometry):
+    """Ellipse.
+
+    Args:
+        center: Center of the ellipse.
+        semimajor_axis: Length of semimajor axis of the ellipse.
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:84]: just semimajor and semiminor

File: deepxde/geometry/geometry_2d.py [hunk 6]
```diff
@@ -65,10 +67,87 @@ def background_points(self, x, dirn, dist2npt, shift):
...
+    """
+
+    def __init__(self, center, semimajor, semiminor, angle=0):
+        self.center = np.array(center, dtype=config.real(np))
+        self.semimajor = semimajor
+        self.semiminor = semiminor
+        self.angle = angle
+        self.c = (semimajor**2 - semiminor**2) ** 0.5
+
+        self.left_center = np.array(
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:97]: what is left_center? left focus? Use the right terminology, see https://en.wikipedia.org/wiki/Ellipse

File: deepxde/geometry/geometry_2d.py [hunk 7]
```diff
@@ -65,10 +67,86 @@ def background_points(self, x, dirn, dist2npt, shift):
...
+        d2 = np.linalg.norm(x - self.focus2, axis=-1)
+        return np.isclose(d1 + d2, 2 * self.semimajor)
+
+    def inside(self, x):
+        d1 = np.linalg.norm(x - self.focus1, axis=-1)
+        d2 = np.linalg.norm(x - self.focus2, axis=-1)
+        return d1 + d2 <= 2 * self.semimajor
+
+    def random_points(self, n, random="pseudo"):
+        """http://mathworld.wolfram.com/DiskPointPicking.html"""
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:129]: `# http://mathworld.wolfram.com/DiskPointPicking.html`

File: deepxde/geometry/geometry_2d.py [hunk 8]
```diff
@@ -65,10 +67,86 @@ def background_points(self, x, dirn, dist2npt, shift):
...
+        rng = sample(n, 2, random)
+        r, theta = rng[:, 0], 2 * np.pi * rng[:, 1]
+        x, y = self.semimajor * np.cos(theta), self.semiminor * np.sin(theta)
+        X = np.sqrt(r) * np.vstack((x, y))
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:133]: Did you check if this algorithm is correct? The link is only for disk, not ellipse.

File: deepxde/geometry/geometry_2d.py [hunk 9]
```diff
@@ -65,10 +67,86 @@ def background_points(self, x, dirn, dist2npt, shift):
...
+        return np.matmul(self.rotation_mat, X).T + self.center
+
+    def uniform_boundary_points(self, n):
+        theta = np.linspace(0, 2 * np.pi, num=n, endpoint=False)
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:137]: Did you check the sampled points?

File: deepxde/geometry/geometry_2d.py [hunk 10]
```diff
@@ -65,10 +67,86 @@ def background_points(self, x, dirn, dist2npt, shift):
...
+        X = np.vstack(
+            (self.semimajor * np.cos(theta), self.semiminor * np.sin(theta))
+        ).T
+        return np.matmul(self.rotation_mat, X.T).T + self.center
+
+    def random_boundary_points(self, n, random="pseudo"):
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:143]: Also this.

File: deepxde/geometry/geometry_2d.py (hunk 8 above)
Comment (anranjiao) [reply deepxde/geometry/geometry_2d.py:133]: That's not exactly random since it's rescaled. I will modify these.

Comment (anranjiao) [reply deepxde/geometry/geometry_2d.py:164]: This random_points() is correct. I modified the other two.

File: deepxde/geometry/geometry_2d.py [hunk 11]
```diff
@@ -65,10 +67,120 @@ def background_points(self, x, dirn, dist2npt, shift):
...
 
+class Ellipse(Geometry):
+    """Ellipse.
+
+    Args:
+        center: Center of the ellipse.
+        semimajor: Semimajor of the ellipse.
+        semiminor: Semiminor of the ellipse.
+        angle: Rotation angle of the ellipse. A positive angle rotates the ellipse
+            clockwise about the center and a negative angle rotates the
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:87]: This line seems short?

Comment (anranjiao) [reply deepxde/geometry/geometry_2d.py:87]: If adding another word, it will exceed 89

File: deepxde/geometry/geometry_2d.py [hunk 12]
```diff
@@ -65,10 +67,120 @@ def background_points(self, x, dirn, dist2npt, shift):
...
+        return theta, cumulative_distance, c
+
+    def _theta_from_arc_length_constructor(self):
+        """Constructs a function that returns the angle associated with a given
+        cumulative arc length for given ellipse.
+        """
+        theta, cumulative_distance, total_arc = self._ellipse_arc()
+        # Construct the inverse arc length function
+        def f(s):
+            assert np.all(s <= total_arc), "s out of range"
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:157]: Delete assert. This is always guaranteed to be true, right?

**Pull request merged**